src/mbgl/gl/offscreen_texture.cpp

```diff
@@ -0,0 +1,96 @@
+#include <mbgl/gl/offscreen_texture.hpp>
+#include <mbgl/gl/renderable_resource.hpp>
+#include <mbgl/gl/context.hpp>
+#include <mbgl/gl/framebuffer.hpp>
+
+namespace mbgl {
+namespace gl {
+
+class OffscreenTextureResource final : public gl::RenderableResource {
+public:
+    OffscreenTextureResource(gl::Context& context_,
+                             const Size size_,
+                             const gfx::TextureChannelDataType type_)
+        : context(context_), size(size_), type(type_) {
+        assert(!size.isEmpty());
+    }
+
+    OffscreenTextureResource(gl::Context& context_,
+                             const Size size_,
+                             gfx::Renderbuffer<gfx::RenderbufferPixelType::Depth>& depth_,
+                             const gfx::TextureChannelDataType type_)
+        : context(context_), size(size_), depth(&depth_), type(type_) {
+        assert(!size.isEmpty());
+    }
+
+    void bind() override {
+        if (!framebuffer) {
+            texture = context.createTexture(size, gfx::TexturePixelType::RGBA, type);
+            if (depth) {
+                framebuffer = context.createFramebuffer(*texture, *depth);
+            } else {
+                framebuffer = context.createFramebuffer(*texture);
+            }
+        } else {
+            context.bindFramebuffer = framebuffer->framebuffer;
+        }
+
+        context.activeTextureUnit = 0;
+        context.scissorTest = false;
+        context.viewport = { 0, 0, size };
+    }
+
+    PremultipliedImage readStillImage() {
+        assert(framebuffer);
+        context.bindFramebuffer = framebuffer->framebuffer;
+        return context.readFramebuffer<PremultipliedImage>(size);
+    }
+
+    gfx::Texture& getTexture() {
+        assert(texture);
+        return *texture;
+    }
+
+private:
+    gl::Context& context;
+    const Size size;
+    optional<gfx::Texture> texture;
+    gfx::Renderbuffer<gfx::RenderbufferPixelType::Depth>* depth = nullptr;
+    const gfx::TextureChannelDataType type;
+    optional<gl::Framebuffer> framebuffer;
+};
+
+OffscreenTexture::OffscreenTexture(gl::Context& context,
+                                   const Size size_,
+                                   const gfx::TextureChannelDataType type)
+    : gfx::OffscreenTexture(size, std::make_unique<OffscreenTextureResource>(context, size_, type)) {
+}
+
+OffscreenTexture::OffscreenTexture(
+    gl::Context& context,
+    const Size size_,
+    gfx::Renderbuffer<gfx::RenderbufferPixelType::Depth>& renderbuffer,
+    const gfx::TextureChannelDataType type)
+    : gfx::OffscreenTexture(
+          size, std::make_unique<OffscreenTextureResource>(context, size_, renderbuffer, type)) {
+}
+
+bool OffscreenTexture::isRenderable() {
```

Comment:
nit: Would `bind` be a better name?

Reply:
I named it `isRenderable` on purpose because the intention of the function isn't necessarily to bind the texture, but to verify that it can be rendered to. In OpenGL, this is the same as trying to bind it, but it's not the same underlying concept.